File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.java
class Solution {

    public static int lowerBound(int target, ArrayList<Integer> al){
        int ans=al.size();
        int low = 0, high = al.size()-1;
        while(low<=high){
            int mid = (low+high)/2;

            if(al.get(mid) >= target){
                ans = mid;
                high = mid-1;
            }else{
                low = mid+1;
            }
        }

        return ans;
    }

    public int lengthOfLIS(int[] nums) {
        
        int n = nums.length;
        int ans=1;
        ArrayList<Integer> al = new ArrayList<Integer>();
        al.add(nums[0]);

        for(int i=1; i<n; i++){
            if(nums[i] > al.get(al.size()-1)){
                al.add(nums[i]);
                ans++;
            }else{
                int ind = lowerBound(nums[i], al);
                al.set(ind, nums[i]);
            }
        }

        return ans;
    }
}

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.cpp
class Solution {
public:

    int lowerBound(vector<int>&contains, int target){
        int ans = contains.size();
        int low=0, high=ans-1;

        while(low<=high){
            int mid = (low+high)/2;
            if(contains[mid] >= target){
                ans = mid;
                high = mid-1;
            }else{
                low = mid+1;
            }
        }

        return ans;
    }

    int lengthOfLIS(vector<int>& nums) {
        
        vector<int> contains;
        contains.push_back(nums[0]);
        int length = 1;
        for(int i=1; i<nums.size(); i++){

            if(nums[i] > contains.back()){
                contains.push_back(nums[i]);
                length++;
            }else{
                int ind = lowerBound(contains, nums[i]);
                contains[ind] = nums[i];
            }
        }

        return length;
    }
};